PerimeterXModule/Internals/PxContext.cs

```diff
@@ -7,167 +7,257 @@
-
-        private string ExtractHttpVersion(HttpContext context)
-        {
-            string serverProtocol = context.Request.ServerVariables["SERVER_PROTOCOL"];
-            if (serverProtocol != null)
-            {
-                int i = serverProtocol.IndexOf("/");
-                if (i != -1)
-                {
-                    return serverProtocol.Substring(i + 1);
-                }
-            }
-            return serverProtocol;
-        }
-
-        public string GetPxCookie()
-        {
-            if (PxCookies.Count == 0)
-            {
-                return null;
-            }
-            return PxCookies.ContainsKey(PxConstants.COOKIE_V3_PREFIX) ? PxCookies[PxConstants.COOKIE_V3_PREFIX] : PxCookies[PxConstants.COOKIE_V1_PREFIX];
-        }
-
-        public Dictionary<string, string> GetHeadersAsDictionary()
-        {
-            Dictionary<string, string> headersDictionary = new Dictionary<string, string>();
-
-            if (Headers != null && Headers.Count() > 0)
-            {
-                headersDictionary = Headers.ToDictionary(header => header.Name, header => header.Value);
-            }
-
-            return headersDictionary;
-        }
-    }
+	public class PxContext
+	{
+		readonly char[] MOBILE_DELIMITER = new char[] { ':' };
+
+		public Dictionary<string, string> PxCookies { get; set; }
+		public object DecodedPxCookie { get; set; }
+		public string PxCookieHmac { get; set; }
+		public string PxCaptcha { get; set; }
+		public string Ip { get; set; }
+		public string HttpVersion { get; set; }
+		public string HttpMethod { get; set; }
+		public List<RiskRequestHeader> Headers { get; set; }
+		public string Hostname { get; set; }
+		public string Uri { get; set; }
+		public string UserAgent { get; set; }
+		public string FullUrl { get; set; }
+		public string S2SCallReason { get; set; }
+		public int Score { get; set; }
+		public string Vid { get; set; }
+		public string UUID { get; set; }
+		public BlockReasonEnum BlockReason { get; set; }
+		public bool MadeS2SCallReason { get; set; }
+		public string S2SHttpErrorMessage { get; set; }
+		public string BlockAction { get; set; }
+		public string BlockData { get; set; }
+		public HttpContext ApplicationContext { get; private set; }
+		public bool SensitiveRoute { get; set; }
+		public PassReasonEnum PassReason { get; set; }
+		public long RiskRoundtripTime { get; set; }
+		public CookieOrigin CookieOrigin { get; set; }
+		public Dictionary<string,string> OriginalTokens { get; set; }
+		public IPxCookie OriginalToken { get; set; }
+		public string OriginalTokenError { get; set; }
+		public string OriginalUUID { get; set; }
+		public object DecodedOriginalToken { get; set; }
+
+		public PxContext(HttpContext context, PxModuleConfigurationSection pxConfiguration)
+		{
+			ApplicationContext = context;
+
+			CookieOrigin = CookieOrigin.COOKIE;
+			PxCookies = new Dictionary<string, string>();
+			OriginalTokens = new Dictionary<string, string>();
```

Comment:
Is there actually a case where there's more than one originalCookie? From the logic it seems like we handle only a single (potential) case. Or do you use it only to be able to use the `PxCookieUtils.BuildCookie` method that expects a dictionary?

Reply:
PxCookieUtils.BuildCookie expects a dict, i did not want to invent new logic for this as this module is already robust in its content